paychmgr/simple.go

```diff
@@ -17,7 +22,211 @@ import (
 	"github.com/filecoin-project/lotus/chain/types"
 )
 
-func (pm *Manager) createPaych(ctx context.Context, from, to address.Address, amt types.BigInt) (cid.Cid, error) {
+type paychApi interface {
+	StateWaitMsg(ctx context.Context, msg cid.Cid, confidence uint64) (*api.MsgLookup, error)
+	MpoolPushMessage(ctx context.Context, msg *types.Message) (*types.SignedMessage, error)
+}
+
+// paychFundsRes is the response to a create channel or add funds request
+type paychFundsRes struct {
+	channel address.Address
+	mcid    cid.Cid
+	err     error
+}
+
+type onCompleteFn func(*paychFundsRes)
+
+// fundsReq is a request to create a channel or add funds to a channel
+type fundsReq struct {
+	ctx        context.Context
+	from       address.Address
+	to         address.Address
+	amt        types.BigInt
+	onComplete onCompleteFn
+}
+
+// getPaych ensures that a channel exists between the from and to addresses,
+// and adds the given amount of funds.
+// If the channel does not exist a create channel message is sent and the
+// message CID is returned.
+// If the channel does exist an add funds message is sent and both the channel
+// address and message CID are returned.
+// If there is an in progress operation (create channel / add funds), getPaych
+// blocks until the previous operation completes, then returns both the channel
+// address and the CID of the new add funds message.
+// If an operation returns an error, subsequent waiting operations will still
+// be attempted.
+func (ca *channelAccessor) getPaych(ctx context.Context, from, to address.Address, amt types.BigInt) (address.Address, cid.Cid, error) {
```

Comment:
can I check my understanding on a couple of sequences, based on my read of this code:

I call getPaych between Alice & Bob with $100 -> will return no address, createCID, nil
I immediate call getPaych between Alice & Bob with $100 again -> will block untill creation is done, return address, addCID, nil
I immediately call addVoucher w/ the address returned from call two -- will still work immediately. I only have to wait for AddCid if I care about the second $100 being in the channel.

Reply:
That's correct 👍